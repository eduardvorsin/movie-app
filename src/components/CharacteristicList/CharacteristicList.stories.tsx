import type { Meta, StoryObj } from '@storybook/react';
import CharacteristicList from './CharacteristicList';

const meta: Meta<typeof CharacteristicList> = {
	title: 'components/CharacteristicList',
	tags: ['autodocs'],
	argTypes: {
		className: {
			description: 'Additional classes for the CharacteristicList'
		},
		testId: {
			description: 'Id for testing the component',
		},
		data: {
			description: 'the data array to be displayed as a list',
		},
	},
	component: CharacteristicList,
	parameters: {
		docs: {
			description: {
				component: 'The component that renders the list in the following format: name - value',
			},
		},
	},
}

export default meta;
type Story = StoryObj<typeof CharacteristicList>;


const data: Record<'en' | 'ru', Record<'name' | 'value', string>[]> = {
	en: [
		{ name: 'budget', value: '1.000.000$' },
		{ name: 'country', value: 'USA' },
		{ name: 'release date', value: '15.12.2023' },
		{ name: 'runtime', value: '2:13' },
		{ name: 'original language', value: 'english' },
		{ name: 'status', value: 'releases' },
	],
	ru: [
		{ name: 'бюджет', value: '1.000.000$' },
		{ name: 'страна', value: 'США' },
		{ name: 'дата релиза', value: '15.12.2023' },
		{ name: 'продолжительность', value: '2:13' },
		{ name: 'язык оригинала', value: 'английский' },
		{ name: 'статус', value: 'релиз' },
	],
};

export const Default: Story = {
	render: (_, context) => {
		const { locale } = context.globals;
		console.log('locale', locale);
		return (
			<CharacteristicList
				data={locale === 'en' ? data.en : data.ru}
			/>
		);
	}
};