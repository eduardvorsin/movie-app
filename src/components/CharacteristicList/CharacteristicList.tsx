import { GeneralProps } from "@/types/shared";

type Props = {
	data: Record<'name' | 'value', string>[],
} & GeneralProps;

export default function CharacteristicList({
	className,
	data,
	testId,
	...props
}: Props) {
	const characteristicListClasses = [
		className,
	].join(' ');

	return (
		<ul
			className={characteristicListClasses}
			data-testid={testId}
			{...props}
		>
			{data.map((characteristic) => (
				<li
					key={characteristic.name}
					className='text-100 flex break-all py-1 xs:py-2'
				>
					<span
						className='text-neutral-700 dark:text-dark-neutral-700 basis-[8rem] xs:basis-[11rem] shrink-0 grow-0 mr-2 transition-colors duration-150'
					>
						{characteristic.name}:
					</span>
					<span
						className='text-neutral-1000 dark:text-dark-neutral-900 flex-grow transition-colors duration-150'
					>
						{characteristic.value}
					</span>
				</li>
			))}
		</ul>
	);
};