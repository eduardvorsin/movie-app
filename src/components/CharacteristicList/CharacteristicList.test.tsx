import { render, screen } from '@testing-library/react';
import CharacteristicList from './CharacteristicList';

const mockData = [
	{
		name: 'first',
		value: 'banana'
	},
	{
		name: 'second',
		value: 'orange'
	},
	{
		name: 'third',
		value: 'apple',
	},
];

describe('CharacteristicList tests', () => {
	it('is rendered correctly', () => {
		render(
			<CharacteristicList
				data={mockData}
				testId='characteristic-list'
			/>);

		expect(screen.getByRole<HTMLUListElement>('list')).toBeInTheDocument();
	});

	it('if the data array is empty then it renders an empty list', () => {
		render(
			<CharacteristicList
				data={[]}
				testId='characteristic-list'
			/>);

		expect(screen.queryByRole<HTMLLIElement>('listitem')).not.toBeInTheDocument();
	});

	it('is is a basic snapshot', () => {
		render(
			<CharacteristicList
				data={mockData}
				testId='characteristic-list'
			/>);

		expect(screen.getByRole<HTMLUListElement>('list')).toMatchSnapshot();
	});

	it('snapshot with an empty data array', () => {
		render(
			<CharacteristicList
				data={[]}
				testId='characteristic-list'
			/>);

		expect(screen.getByRole<HTMLUListElement>('list')).toMatchSnapshot();
	});
});