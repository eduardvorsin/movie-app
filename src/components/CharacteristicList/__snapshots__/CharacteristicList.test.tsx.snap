// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacteristicList tests is is a basic snapshot 1`] = `
<ul
  class=""
  data-testid="characteristic-list"
>
  <li
    class="text-100 flex break-all py-1 xs:py-2"
  >
    <span
      class="text-neutral-700 dark:text-dark-neutral-700 basis-[8rem] xs:basis-[11rem] shrink-0 grow-0 mr-2 transition-colors duration-150"
    >
      first
      :
    </span>
    <span
      class="text-neutral-1000 dark:text-dark-neutral-900 flex-grow transition-colors duration-150"
    >
      banana
    </span>
  </li>
  <li
    class="text-100 flex break-all py-1 xs:py-2"
  >
    <span
      class="text-neutral-700 dark:text-dark-neutral-700 basis-[8rem] xs:basis-[11rem] shrink-0 grow-0 mr-2 transition-colors duration-150"
    >
      second
      :
    </span>
    <span
      class="text-neutral-1000 dark:text-dark-neutral-900 flex-grow transition-colors duration-150"
    >
      orange
    </span>
  </li>
  <li
    class="text-100 flex break-all py-1 xs:py-2"
  >
    <span
      class="text-neutral-700 dark:text-dark-neutral-700 basis-[8rem] xs:basis-[11rem] shrink-0 grow-0 mr-2 transition-colors duration-150"
    >
      third
      :
    </span>
    <span
      class="text-neutral-1000 dark:text-dark-neutral-900 flex-grow transition-colors duration-150"
    >
      apple
    </span>
  </li>
</ul>
`;

exports[`CharacteristicList tests snapshot with an empty data array 1`] = `
<ul
  class=""
  data-testid="characteristic-list"
/>
`;
